### ISSUE TITLE
[HIGH] Hold doesn't show in three dot menu until report is opened

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#52401]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.60-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @garrettmknight 
**Slack conversation** (hyperlinked to channel name): [ts_external_expensify_expense](https://expensify.slack.com/archives/C06ML6X0W9L/p1731403866211499)

## Action Performed:
1. Create a workspace
2. Invite Submitter + Approver A
3. Enable Workflows > Approval + Delay Submission
4. Set Approver A as approver
5. As Submitter, create expense in WS chat + submit
6. Tag the approver in the workspace chat and invite them
7. As Approver A, navigate to Submitter's WS chat + click the three dot menu on the expense

## Expected Result:
Since this is a one expense report, you should see Hold in the three dot menu for the report
## Actual Result:
The Hold action is not available in the menu until you open the expense and go back

## Workaround:
You don't see Hold until you access the expense/report and back out into the WS chat again.
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details>

https://github.com/user-attachments/assets/1b863b87-85c6-46e0-8a12-7e11596fc678

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021857521697236720139</li>
        <li>Upwork Job ID: 1857521697236720139</li>
        <li>Last Price Increase: 2024-11-22</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @neil-marcellini</details>

**Comment by CortneyOfstad** [comment]
I was able to recreate this under my expensifail.com account, so getting some eyes on this!

**Comment by me-ZaidAli** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The three dot menu doesn't show `Hold` menu item when a new report is created

### What is the root cause of that problem?
The issue is in the file `BaseReportActionContextMenu` file. This is the piece of code where the problem is. 
```Javascript
  const requestParentReportAction = useMemo(() => {
        if (isMoneyRequestReport || isInvoiceReport) {
            if (!paginatedReportActions || !transactionThreadReport?.parentReportActionID) {
                return undefined;
            }
            return paginatedReportActions.find((action) => action.reportActionID === transactionThreadReport.parentReportActionID);
        }
        return parentReportAction;
    }, [parentReportAction, isMoneyRequestReport, isInvoiceReport, paginatedReportActions, transactionThreadReport?.parentReportActionID]);

    const moneyRequestAction = transactionThreadReportID ? requestParentReportAction : parentReportAction;
```

The `moneyRequestAction` is required to be true to show `Hold` menu item. We have `transactionThreadReportID` and so the first operation of the ternary operator gets evaluated. `requestParentReportAction` returns undefined since this condition gets evaluated to true due to absence of `parentReportActionID` in `transactionThreadReport`.
```javascript
if (!paginatedReportActions || !transactionThreadReport?.parentReportActionID) {
                return undefined;
}
```
like mentioned in the workaround, once we open the report and get back to the chat, we start seeing `Hold` coz `transactionThreadReport` get populated with all it's fields including `parentReportActionID` and 
```javascript
return paginatedReportActions.find((action) => action.reportActionID === transactionThreadReport.parentReportActionID);
```

gets evaluated to true.

### What changes do you think we should make in order to solve the problem?
The problem is with report loading since our condition relies on `transactionThreadReport` and when the report is first created, `transactionThreadReport` object is not loaded fully (missing major fields including `parentReportActionID`). So have got to make sure it gets loaded properly from Onyx.


### What alternative solutions did you explore? (Optional)
Alternate would be to change the condition required to show `Hold` but I wouldn't go that route as it might break some other stuff that we are unaware of.

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by me-ZaidAli** [comment]
Contributor details
Your Expensify account email: zaid.ali5430@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01dc2d382489d84c0b

**Comment by ahmedGaber93** [comment]
Sorry for the delay, Reviewing today

**Comment by ahmedGaber93** [comment]
I think this is a BE issue.

The [reportAction](https://github.com/Expensify/App/blob/9368566b52364a34747d8f97d3de7f427b3e0712/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L135) data that comes from `pusher` doesn't have [childReportID](url) that used [here](https://github.com/Expensify/App/blob/9368566b52364a34747d8f97d3de7f427b3e0712/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L146) which break showing the hold item. And if you open any other report, then go back to the expense report to load the `API` data, you will see the issue is fixed.

https://github.com/user-attachments/assets/f6851eb4-b66e-4cbf-8806-fca1fa4fc0a0

<br>

Comparison of `reportAction` between `pusher` and `API` data.

![Screenshot 2024-11-19 at 10 01 23 PM](https://github.com/user-attachments/assets/b6e2e0de-9476-4c94-911b-9eda3866e2d9)


1. add console.log for [reportAction](https://github.com/Expensify/App/blob/9368566b52364a34747d8f97d3de7f427b3e0712/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L135)
2. reproduce the issue
3. hover on the `reportAction` view to log the `reportAction` data that come from the `pusher` (doesn't have `childReportID`)
4. open any other report, then back to the expense report
5. hover on the `reportAction` view to log the `reportAction` data that come from the `API` (have the `childReportID`)

**Comment by ahmedGaber93** [comment]
@CortneyOfstad I think this is a BE issue and should be internally.

**Comment by me-ZaidAli** [comment]
@ahmedGaber93 which environment are you testing it on? for me on staging the `childReportId` is present in `reportAction` . The issue that you mentioned with `canHoldUnholdReportAction` happens [here](https://github.com/Expensify/App/blob/7bfb34527508b47732cab6d4280a943ffab48066/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L155) since we are unable to fetch `childReport` for the `IOU` report action. So when we open the report, FE fetches the report data from BE and it's cached locally for us to access in the chat. `transactionThreadReport` becomes available and so we start seeing the `Hold` option.
In addition to this issue there is a lot of refactoring needed in the component since it's filled with stuff and conditions that are not even needed.

**Comment by ahmedGaber93** [comment]
@me-ZaidAli Testing on DEV with latest main branch.

**Comment by ahmedGaber93** [comment]
> The issue that you mentioned with canHoldUnholdReportAction happens [here](https://github.com/Expensify/App/blob/7bfb34527508b47732cab6d4280a943ffab48066/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L155) since we are unable to fetch childReport for the IOU report action.

I agree with this part, the `parentReportActionID` also is not present on `transactionThreadReport` that come from `pusher`, so we also need to fix it from BE side

![Screenshot 2024-11-20 at 6 15 39 AM](https://github.com/user-attachments/assets/f41572b7-b9cf-4b98-a139-d3728a20226b)

**Comment by ahmedGaber93** [comment]
@CortneyOfstad bump https://github.com/Expensify/App/issues/52401#issuecomment-2487676060

**Comment by muttmuure** [comment]
Moving to Hot picks

**Comment by ahmedGaber93** [comment]
Not overdue, waiting for an internal developer.

**Comment by ahmedGaber93** [comment]
Not overdue, waiting for an internal developer.

**Comment by tgolen** [comment]
Can someone please open a separate GH issue for an internal engineer using [this template](https://github.com/Expensify/App/issues/new?assignees=&labels=Daily%2C+Internal%2C+AutoAssignerNewDotQuality&projects=&template=Internal.md)? It will ensure that an engineer is assigned and they have the full context of what they need to do without reading through the entire comment history and try to piece together what the problem is.

**Comment by neil-marcellini** [comment]
I'll grab it and sort it out

**Comment by CortneyOfstad** [comment]
Sounds good @tgolen and thank you @neil-marcellini!

**Comment by neil-marcellini** [comment]
I wasn't able to reproduce this on the latest dev environment. I tried to reproduce on staging, but I'm blocked by a separate bug. I will report it, put this one on HOLD, and try to fix that bug first.

**Comment by neil-marcellini** [comment]
This was reported in [Slack here](https://expensify.slack.com/archives/C06ML6X0W9L/p1733313744201329), we'll discuss from there.

**Comment by CortneyOfstad** [comment]
Latest update from Slack post [here](https://expensify.slack.com/archives/C06ML6X0W9L/p1733514984515659?thread_ts=1733313744.201329&cid=C06ML6X0W9L)

**Comment by neil-marcellini** [comment]
I have a fix in review for the issue that was blocking this. [402 Approvals can only be set for members of the policy” error when setting approver for new user added to workspace](https://github.com/Expensify/App/issues/53738). I'm going to put this on HOLD until that's deployed, then we can check if this issue is still occurring on staging.

**Comment by ahmedGaber93** [comment]
I have unassigned me because this issue doesn't need C+, but I will be happy to help if needed.

**Comment by CortneyOfstad** [comment]
Sounds good @neil-marcellini and thank you @ahmedGaber93!

**Comment by neil-marcellini** [comment]
I'm able to reproduce now, but the steps are a little different. In fact, the steps make it quite a rare bug so I think we should lower the priority and maybe move it to a different project. You have to tag the approver and invite them to your workspace chat to reproduce the issue.

Bug with new steps

https://github.com/user-attachments/assets/1b863b87-85c6-46e0-8a12-7e11596fc678

Bug with the old steps
- Notice how you never see the user's workspace chat, nor the report preview. I can't manage to pull up the user's workspace chat via search either.

https://github.com/user-attachments/assets/08a4c2cc-e8a5-4ab4-b753-082b1f0c937b

New steps
- I have updated the description
1. Create a workspace
2. Invite Submitter + Approver A
3. Enable Workflows > Approval + Delay Submission
4. Set Approver A as approver
5. As Submitter, create expense in WS chat + submit
6. Tag the approver in the workspace chat and invite them
7. As Approver A, navigate to Submitter's WS chat + click the three dot menu on the expense

**Comment by neil-marcellini** [comment]
Not overdue

**Comment by neil-marcellini** [comment]
Closing as explained in Slack internally [here](https://expensify.slack.com/archives/C03U7DCU4/p1734051243790959). For external folks looking, I think it's not worth the effort to fix since it's now a rare edge case.

**Comment by mountiny** [comment]
Since Neil is ooo for the rest of the week, marking it as a hot pick and he can pick it up again once he is back if he wants to unless anyone else jumped on it

**Comment by CortneyOfstad** [comment]
Thanks @mountiny!

**Comment by CortneyOfstad** [comment]
Just a heads up that I will be OOO starting this afternoon (December 20th) and will be returning January 6th. If any action is needed from me during that time, please feel free to reapply the BZ label, otherwise I will address it as soon as I am back in office.

Thank you!

**Comment by CortneyOfstad** [comment]
Hey @neil-marcellini! I see you're back in office! Just checking to see if this is something you were still interested in — thanks!

**Comment by CortneyOfstad** [comment]
Bump @neil-marcellini on the comment above ^^^ thanks!

**Comment by CortneyOfstad** [comment]
Still a hot pick, bug going to see if this can consistently be recreated 👍

**Comment by CortneyOfstad** [comment]
Still waiting for this to be tested

**Comment by garrettmknight** [comment]
We can retest, but this is still happening to me often.

**Comment by CortneyOfstad** [comment]
@neil-marcellini I see you originally did a PR that was blocking this based on the comment [here](https://github.com/Expensify/App/issues/52401#issuecomment-2529296176). You explored fixing this but after a Slack discussion [here](https://expensify.slack.com/archives/C03U7DCU4/p1734051243790959), it was decided to not move forward, only for the GH to be reopened. 

Do you have the bandwidth to take this on? If not, please let me know and I can ping the Hot Picks team to see if someone can take a look and get this resolved. Thanks!

**Comment by garrettmknight** [comment]
@CortneyOfstad thanks for keeping on this one. Just hit me again today.

**Comment by mallenexpensify** [comment]
Added `AutoAssignerNewDotQuality` to assign an engineer, I didn't want to _assume_ that Neil should be assigned cuz he worked on a related PR.  @neil-marcellini and @iwiznia , 👀 plz so we can move this forward.  It's been updated to CRITICAL on #quality (hence the auto-assignment Ioni).  Thx.   Thx @CortneyOfstad for bumping up the status too.

**Comment by iwiznia** [comment]
> The [reportAction](https://github.com/Expensify/App/blob/9368566b52364a34747d8f97d3de7f427b3e0712/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L135) data that comes from pusher doesn't have childReportID that used [here](https://github.com/Expensify/App/blob/9368566b52364a34747d8f97d3de7f427b3e0712/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L146) which break showing the hold item. And if you open any other report, then go back to the expense report to load the API data, you will see the issue is fixed.

@ahmedGaber93 what API command(s) are you talking about here?

**Comment by ahmedGaber93** [comment]
> what API command(s) are you talking about here?

@iwiznia Taking about `api/OpenReport` for the expense chat.

When the approver keeping the expense chat open, and receive a request money, the `reportAction` data will be received by the pusher, and it will not have `childReportID`.  

And if the user opens any other report, then reopening the expense chat again (API `OpenReport` will load the expense report again with all its reportActions), in this time the request money preview `reportAction` will come with the `childReportID` we need it.

So we need to back `childReportID` in the pusher data.
___

Why we need `childReportID` to show hold option?

We use it [here](https://github.com/Expensify/App/blob/9368566b52364a34747d8f97d3de7f427b3e0712/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L146-L170), to get `moneyRequestAction` which used in [canHoldUnholdReportAction](https://github.com/Expensify/App/blob/9368566b52364a34747d8f97d3de7f427b3e0712/src/libs/ReportUtils.ts#L3286)

**Comment by CortneyOfstad** [comment]
@iwiznia and @ahmedGaber93 any updates on this? If there is any action needed from me, just let me know!

**Comment by ahmedGaber93** [comment]
@iwiznia Sorry I haven't tested it out recently. When testing it on Nov 2024, this was the case with proven video here https://github.com/Expensify/App/issues/52401#issuecomment-2488621537, but now I agree `childReportID` is already exist in pusher data.

___

The case now looks missing `transactionThreadReport?.parentReportActionID` in pusher data that used [here](https://github.com/Expensify/App/blob/ad7462d095158adab111ff34e86f39b00d7222b1/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L183), and here are the steps I follow to catch it.

1. Add console.log [here](https://github.com/Expensify/App/blob/7bfb34527508b47732cab6d4280a943ffab48066/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L159) to log `transactionThreadReportID` `console.log('transactionThreadReportID', transactionThreadReportID);`
2. Approver keeping the expense chat open
3. Submitter request money request
4. Approver receive the request 
5. Click on three dots, there is no "hold" option
6. Get `transactionThreadReportID` that logged in the console.
7. Open OnyxDB and search for report_ `transactionThreadReportID`, the report just have `participants` and missed `parentReportActionID` that used [here](https://github.com/Expensify/App/blob/ad7462d095158adab111ff34e86f39b00d7222b1/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L183)
8. open any other report from LHN then back again to user's expense report (wait a second until API `OpenReport` complete)
8. refresh the ONYX database and look again for the report_ `transactionThreadReportID`, the report now load completely on OnyxDB
9. Click on three dots, the "hold" option is shown now

https://github.com/user-attachments/assets/b6034a83-1851-435a-b90b-535f9d5de343

**Comment by iwiznia** [comment]
> @iwiznia Sorry I haven't tested it out recently. When testing it on Nov 2024, this was the case with proven video here https://github.com/Expensify/App/issues/52401#issuecomment-2488621537, but now I agree childReportID is already exist in pusher data.

Ah all good, maybe things changed since then.

> The case now looks missing transactionThreadReport?.parentReportActionID in pusher data that used [here](https://github.com/Expensify/App/blob/ad7462d095158adab111ff34e86f39b00d7222b1/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L183), and here are the steps I follow to catch it.

Are you saying that what I said above is wrong and the problem is not that the child report is not in onxy?

**Comment by ahmedGaber93** [comment]
The child report exists in onyx, the problem now is the  transactionThreadReport?.parentReportActionID is not exist.

**Comment by iwiznia** [comment]
Hmmmm are there 2 cases where the bug exists? Because I checked yesterday and what I said above was true.

**Comment by iwiznia** [comment]
ok I retested my case just to be sure and for sure that's one of the problems.

**Comment by iwiznia** [comment]
That seems like a different flow, can you write down step by step the actions you took?

**Comment by ahmedGaber93** [comment]
1. Add console.log [here](https://github.com/Expensify/App/blob/7bfb34527508b47732cab6d4280a943ffab48066/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L145) to log `reportAction?.childReportID`
2. Approver keeping the expense chat open
3. Submitter request money request
4. Approver receive the request 
5. Click on three dots, there is no "hold" option
6. Get `reportAction?.childReportID` that logged in the console.
7. Open OnyxDB and search for report_ `reportAction?.childReportID`, see child report exists in OnyxDB.

**Comment by CortneyOfstad** [comment]
@iwiznia — does the flow above work, or does it need further editing?

**Comment by iwiznia** [comment]
I just tested that case and yeah that's also another way of reproducing the error.

**Comment by iwiznia** [comment]
Since it seems there's several cases where data is not loaded and we expect it to be, we are going to add https://github.com/Expensify/App/issues/58499 in order to detect those places.

**Comment by mallenexpensify** [comment]
@iwiznia should we put this issue on hold pending https://github.com/Expensify/App/issues/58499 ?   If so, I'd like to add that issue to `#quality` with the same CRITICAL status as here since we'd be holding on it.

**Comment by iwiznia** [comment]
Sure

**Comment by CortneyOfstad** [comment]
Draft PR is here — https://github.com/Expensify/App/pull/59191. 

Just a heads up that I will be OoO this afternoon (March 28) to April 7 — if action is needed on the BZ side of things, please reapply the BZ label. Otherwise I will review when I get back. Thanks!

**Comment by CortneyOfstad** [comment]
PR is still being worked on 👍

**Comment by mvtglobally** [comment]
Issue is reproducible during KI retests. 

https://github.com/user-attachments/assets/62668280-d363-4d74-befe-cbedf49e4fdb

**Comment by mallenexpensify** [comment]
- https://github.com/Expensify/App/pull/59191#issuecomment-2816150158

Took off hold, the above PR hit production a few days ago.  @iwiznia what's next step here?

**Comment by CortneyOfstad** [comment]
Thanks @mallenexpensify!

**Comment by CortneyOfstad** [comment]
@iwiznia — bump on the comment [here](https://github.com/Expensify/App/issues/52401#issuecomment-2818106483). Thanks!

**Comment by iwiznia** [comment]
I need to look into it again but should not hold anymore

**Comment by mallenexpensify** [comment]
Working on the `#quality` weekly update and posted this
> @iwiznia  this has stalled, can you prioritize working on it?  If expert devs can help, can you please post in [#expert-contributors](https://slack.com/archives/C08CZDJFJ77)?

**Comment by CortneyOfstad** [comment]
@iwiznia is there an update on Matt's comment above? Thanks!

**Comment by iwiznia** [comment]
No update, sorry. I also just realized this was changed to critical, any idea why?

**Comment by mallenexpensify** [comment]
It's been CRITICAL for months, was on hold for a lil while though. 

I'm unsure how often Approvers put reports on hold.  We could _demote_ to HIGH if we think this only affects a "subset of users" 


- CRITICAL - UX Reliability, API Reliability and Performance bugs that are noticeable for a majority of users
- HIGH -  UX Reliability, API Reliability and Performance issues that are part of main flows and affect a subset of users
- MEDIUM - Debugging Tools to help make fixing existing bugs easier and getting fixes to customers faster and Architectural Improvements to the underlying framework and design of the app to improve performance, scalability, and maintainability
- LOW - Not very noticeable UX Reliability, API Reliability or Performance issues that only affect a single user or subset of users

**Comment by garrettmknight** [comment]
I think HIGH works - approvers are a subset of users.

**Comment by CortneyOfstad** [comment]
Hey @iwiznia — checking to see if there is an update? Thanks!

**Comment by iwiznia** [comment]
The only thing I can think of is to move all the logic to the backend and return something like `canHold: bool, canUnhold: bool` directly from the server as part of the expense report's data OR as part of the reportActions's data.

**Comment by iwiznia** [comment]
Or maybe even return a `allowedActions` property with a list of actions that are allowed and basically replace all the `shouldShow` methods of `ContextMenuActions.tsx` with the data coming from the server.

**Comment by iwiznia** [comment]
Conversation about the proposal above here https://expensify.slack.com/archives/C03TQ48KC/p1748001911447839

**Comment by CortneyOfstad** [comment]
Thanks @iwiznia!

**Comment by CortneyOfstad** [comment]
PR has been merged, but has not deployed to staging yet. 

@iwiznia — just a heads up that I am heading OoO this afternoon (6/3) to Monday (6/9) so if any action is needed from a BZ perspective, please reapply the label. Otherwise I will get it addressed as soon as I am back. Thanks!

**Comment by CortneyOfstad** [comment]
I'm back! PR was deployed to staging 4 days ago, so will keep an eye out for when it is in production 👍

**Comment by CortneyOfstad** [comment]
PR was merged and  issue is resolved (unable to recreate) — closing this out!
